content/realtime/delta.textile

```diff
@@ -0,0 +1,222 @@
+---
+title: Delta mode subscription
+section: realtime
+index: 32
+jump_to:
+  Help with:
+    - Overview#overview
+    - Examples#examples
+  Parameters:
+    - delta#delta
+---
+
+h2(#overview). Overview
+
+Often a channel carries messages that represent a series updates to a particular object or document and, as such, there is a significant degree of similarily between successive messages. Delta mode is a way for a client to subscribe to a channel so that message payloads sent over the wire contain only the difference (ie the delta) between the present message and the previous message on the channel. The client can then apply the delta to the previous message to obtain the full payload. Using delta mode can signficantly reduce the encoded size of each message in the case that message payloads change by differences that are small relative to the size of the value. This reduction in size can reduce bandwidth costs, reduce transit latencies, and enable greater message throughput on a connection.
+
+The present delta mode implementation supports a single representation of a delta, [VCDIFF](https://tools.ietf.org/html/rfc3284). However, the protocol and API are designed to allow other representations to be used in the future.
+
+Since version 1.2, Ably libraries support delta subscriptions; subscribing in delta mode is enabled for a given channel by specifying a [delta channel param](./channel-params) with the value @vcdiff@. This will cause delta messages to be generated by the server and sent to the client, and the library reconstitutes the original message payload. The end result is that messages on the channel are delivered to the subscriber's listener in just the same way as happens with a normal subscription.
+
+h2(#delta-processing). Delta processing
+
+Deltas apply to the principal payload of a @Message@ published via Ably, which is the @data@ member. Other elements of a message, such as @clientId@, @name@, or @extras@ are unchanged by use of deltas and are not compressed.
+
+Deltas are supported for realtime subscriptions only. Messages retrieved via the history API, and messages delivered to Reactor endpoints, are not compressed. Support for delta-compressed messages via Reactor is under consideration for the future.
+
+Delta compression via @vcdiff@ is supported for all payloads, whether string or binary, or JSON-encoded. The delta algorithm processes message payloads as opaque binaries and has no dependency on the stucture of the payload - it does not process line-oriented diffs, for example. In principle, @vcdiff@ deltas can be applied to encrypted message payloads, but in practice this provides no benefit because there is no similarity between successive encrypted payloads even on identical or near-identical plaintext message payloads.
```

Comment:
Again maybe the encrypted payloads statement is not true because of the delta compression thing.

Reply:
nah, properly encrypted data is incompressible. (In particular we use aes in cbc mode, not ecb; an encrypted payload will be indistinguishable from random data no matter how much repetition is in the plaintext)